### PR TITLE
Clarify storage infrastructure

### PR DESCRIPTION
**Pull request by annevk**
- [ ] At least two implementers are interested (and none opposed):
   * Chrome
   * Firefox
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * N/A (or maybe for sessionStorage implications?)
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * N/A

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/storage/86.html" title="Last updated on May 13, 2020, 12:42 PM UTC (21861db)">Preview</a> | <a href="https://whatpr.org/storage/86/8e9ef53...21861db.html" title="Last updated on May 13, 2020, 12:42 PM UTC (21861db)">Diff</a>

### CODE REVIEW

**Comment by annevk** [comment]
I think this now defines sufficient infrastructure to define storage and session storage APIs. It does not:

1. Define a replacement operation that is needed for #4 and `Clear-Site-Data`.
2. Change the keying away from origin (though it does acknowledge and isolate the problem to a single point).

A storage API would do something like:

> Let _map_ be the result of running obtain a storage bucket area map with _environment_, "storage", and "localStorage".

It can then use that _map_ as it pleases, for instance in response to API calls.

(Writing this down I realize we should probably expose a separate "obtain a storage bucket area map" and a "obtain a session storage bucket area map" so APIs do not have to pass a storage type.)

It should also provide a solid basis for any kind of multiple storage bucket API though in the absence of one looks a bit complicated.

If this looks agreeable I would like, in order:

1. Merge this.
2. Work on defining the replacement operation and get that merged.
3. Work on a PR for localStorage/sessionStorage to make use of this.
4. Continue work on storage partitioning for improvements to the storage key.

I could wait with 1 until 2 is further along, but I would greatly appreciate detailed feedback on this PR first.

**Comment by asutherland** [comment]
I understand this to be the hierarchy:
1. Storage Map
2. Storage Unit
3. Storage Bucket
4. (Storage) Area

While the following sounds very silly, my brain has a lot of trouble with there being no inherent containment relationship among the terms in use and with 3 of the 4 being generic.  What about something like:
1. Storage Shed
2. Storage Shelf
3. Storage Bucket
4. Storage Bottle

**Comment by annevk** [comment]
Note that now it's clearly prefixed, "storage box" would work as well. Nobody commented on that suggestion from https://github.com/whatwg/storage/issues/51#issuecomment-338227477.

**Comment by pwnall** [comment]
I'd like to express mild support for keeping bucket. It wasn't the first word that came to my mind, but it has grown on me.

Storage box is a bit of a mouthful. I worry that developers everywhere (web apps, browsers) will end up with variables named `box` instead of `storageBox`. I'd definitely use the word "boxes" when discussing this concept with my colleagues.

For better or worse, "bucket" is not used in other parts of the web platform (as far as I know). I think this will make everyone's (web developers, browser vendors) life easier when searching in code / docs. I expect that this concept will be a big deal, and deserves a unique name.

**Comment by annevk** [comment]
Fair. Any thoughts on @asutherland's suggested names? (And the remainder of the PR? 😊)

**Comment by annevk** [comment]
cc @youennf

File: storage.bs [hunk 1]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
+a
+<dfn for="abstract storage bucket/area,storage bucket/area,session storage bucket/area">proxy map reference set</dfn>,
+which is initially an empty <a for=/>set</a>.
 
-<p>Each <a for=/>origin</a> has an associated <a>storage unit</a>. A <a>storage unit</a> contains a
-single <dfn export id=bucket oldids=box>bucket</dfn>. [[HTML]]
+<p>How an <a>abstract storage bucket</a>'s <a for="abstract storage bucket">area</a>'s
+<a for="abstract storage bucket/area">map</a> is stored is <a>implementation-defined</a>. How it is
+made available across <a>agent</a> or even <a>agent cluster</a> boundaries is
+<a>implementation-defined</a>.
```
Comment (domenic) [inline storage.bs:235]: This is a bit confusing. Upon re-reading it's clear, but at first I misread it as "_Whether_ it is made available across agent or event agent cluster boundaries is implementation-defined".

I'd rephrase this section as something like

> An area's map is where the actual data meant to be stored lives. User agents are expected to store this data, and make it available across agents and agent clusters, in an implementation-defined manner, so that when this specification (or other specifications?) access the contents of the map, they are available.

File: storage.bs [hunk 2]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
 
+<hr>
 
-<h3 id=buckets oldids=boxes>Buckets</h3>
+<p>A <dfn id=bucket oldids=box>storage bucket</dfn> is an <a>abstract storage bucket</a> for storage
+APIs.
```
Comment (domenic) [inline storage.bs:240]: This is confusing because "session storage APIs" sounds like a subset of "storage APIs". I think it'd be good to have two distinct terms. (Maybe changing the `"storage"` and `"session-storage"` types as well.)

File: storage.bs [hunk 3]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
+ <li><p>Return <var>bucket</var>.
+</ol>
+
+
+<h3 id=storage-proxy-maps>Storage proxy maps</h3>
+
+<p>A <dfn>storage proxy map</dfn> is equivalent to a <a for=/>map</a>, except that all operations
+are instead performed on its <dfn for="storage proxy map">backing map</dfn>.
+
+<p class="note">This allows for the <a for="storage proxy map">backing map</a> to be replaced.
```
Comment (domenic) [inline storage.bs:281]: When is replacing the map necessary? It'd be good to state that to motivate this somewhat-strange architecture.

File: storage.bs [hunk 4]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
+ <a for="storage unit">map</a>["<code>default</code>"].
+
+ <li><p>Let <var>area</var> be <var>bucket</var>'s
+ <a for="storage bucket">map</a>[<var>identifier</var>].
+
+ <li><p>Let <var>proxyMap</var> be a new <a>storage proxy map</a> whose
+ <a for="storage proxy map">backing map</a> is <var>area</var>'s
+ <a for="storage bucket/area">map</a>.
+
+ <li><p>Append a reference to <var>proxyMap</var> to <var>area</var>'s
```
Comment (domenic) [inline storage.bs:323]: Link append?

"a reference to" seems redundant and a bit confusing?

File: storage.bs [hunk 5]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
+ <a for="storage bucket/area">proxy map reference set</a>.
+
+ <li><p>Return <var>proxyMap</var>.
+</ol>
+
+<p>To <dfn export>obtain a storage bucket area map</dfn>, given an
```
Comment (domenic) [inline storage.bs:329]: The above-mentioned intro/model section I suggested should prominently point out that these are the entry points specifications will usually use.

File: storage.bs [hunk 6]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<p>A <dfn>storage type</dfn> is "<code>storage</code>" or "<code>session-storage</code>".
+
+<p>A <dfn>storage key</dfn> is an <a for=/>origin</a>. [[HTML]]
+
+<p class=XXX>This is expected to change, see
+<a href="https://privacycg.github.io/storage-partitioning/">Client-Side Storage Partitioning</a>.
+
+<p>A <dfn id=site-storage-unit>storage unit</dfn> is the least granular unit of storage, other than
+the user agent. It holds a <dfn for="storage unit">map</dfn>, which is a <a for=/>map</a> of
+<a for=/>ASCII strings</a> to <a>abstract storage buckets</a>.
```
[the remark this answers is not in the document]
Comment (annevk) [reply storage.bs:122]: These are internal and not exposed until we have buckets at which point they could be whatever buckets need.

File: storage.bs [hunk 7]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<p>A <dfn export>storage endpoint</dfn> is a storage or session storage API that uses the
+infrastructure defined by this standard to keep track of its storage needs.
+
+<p>A <a>storage endpoint</a> has an <dfn for="storage endpoint">identifier</dfn>, which is a
+<a>storage identifier</a>.
+
+<p>A <a>storage endpoint</a> also has <dfn for="storage endpoint">types</dfn>, which is a
+<a for=/>set</a> of <a>storage types</a>.
+
+<p>A <dfn>storage identifier</dfn> is an <a for=/>ASCII string</a>.
```
[the remark this answers is not in the document]
Comment (annevk) [reply storage.bs:182]: Yeah, the idea is to match whatever we need for quota API. That's ASCII per design principles. Not sure I want to lay out casing rules and such though.

File: storage.bs [hunk 8]
```diff
@@ -94,37 +100,251 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<h3 id=storage-endpoints>Storage endpoints</h3>
+
+<p>A <dfn export>storage endpoint</dfn> is a storage or session storage API that uses the
+infrastructure defined by this standard to keep track of its storage needs.
+
+<p>A <a>storage endpoint</a> has an <dfn for="storage endpoint">identifier</dfn>, which is a
+<a>storage identifier</a>.
+
+<p>A <a>storage endpoint</a> also has <dfn for="storage endpoint">types</dfn>, which is a
+<a for=/>set</a> of <a>storage types</a>.
```
[the remark this answers is not in the document]
Comment (annevk) [reply storage.bs:180]: It's been requested to be able to use APIs with different backends. E.g., session IDB. This would allow for that.

File: storage.bs (hunk 2 above)
Comment (annevk) [reply storage.bs:240]: Let me know if you think of something good.

File: storage.bs (hunk 3 above)
Comment (annevk) [reply storage.bs:281]: Clear-Site-Data and storage access API as discussed in the issue.

File: storage.bs (hunk 6 above)
[the remark this answers is not in the document]
Comment (domenic) [reply storage.bs:122]: It'd be good to state that, and probably relax the ASCII requirement then.

File: storage.bs (hunk 7 above)
[the remark this answers is not in the document]
Comment (domenic) [reply storage.bs:182]: It'd be good to state that, and either remove the ASCII requirement (since it becomes implicit) or state exactly how the character set and casing for these are determined.

File: storage.bs (hunk 8 above)
[the remark this answers is not in the document]
Comment (domenic) [reply storage.bs:180]: It'd be good to state that.

[the remark this answers is not in the document]
Comment (domenic) [reply storage.bs:180]: Or just remove this extra complexity until such features arrive?

File: storage.bs (hunk 2 above)
Comment (domenic) [reply storage.bs:240]: I guess "persistent" has a different meaning. "local", maybe.

File: storage.bs (hunk 3 above)
Comment (domenic) [reply storage.bs:281]: It'd be good to state that.

File: storage.bs (hunk 7 above)
[the remark this answers is not in the document]
Comment (annevk) [reply storage.bs:182]: I think it's generally useful to state the narrowest available type of something. We do that in the URL Standard too. I don't think that means going overboard though.

File: storage.bs (hunk 8 above)
[the remark this answers is not in the document]
Comment (annevk) [reply storage.bs:180]: That's an alternative. I wanted to have something that could address the full set of use cases everyone had in mind. As it doesn't really affect what's observable I think it's fine to leave it in for a bit until those various ideas get more concrete.

It would also be somewhat annoying to have to refactor various specifications when those features arrive, but maybe it's not too bad.

File: storage.bs (hunk 2 above)
Comment (annevk) [reply storage.bs:240]: I did think of that, but calling IDB a local storage API doesn't seem very clear either. Hmm.

Comment (domenic) [reply storage.bs:240]: eduring, durable, lasting, long-lived?

**Comment by annevk** [comment]
I plan to address the remainder of @domenic's comments tomorrow as well as fully adopt @asutherland's fancy naming scheme. If you all have comments beyond those now would be a great time to make them. Thanks everyone!

File: storage.bs [hunk 9]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<p>A <dfn>storage key</dfn> is an <a for=/>origin</a>. [[HTML]]
+
+<p class=XXX>This is expected to change, see
+<a href="https://privacycg.github.io/storage-partitioning/">Client-Side Storage Partitioning</a>.
+
+<p>A <dfn id=site-storage-unit>storage unit</dfn> is the least granular unit of storage, other than
+the user agent. It holds a <dfn>bucket map</dfn>, which is a <a for=/>map</a> of
+<a for=/>strings</a> to <a>storage buckets</a>.
+
+<p class="note">For now "<code>default</code>" is the only <a for=map>key</a> that exists. See
```
Comment (inexorabletash) [inline storage.bs:128]: Maybe clarify "that exists" to "that exists for bucket maps" ? Just because many keys/maps are defined nearby.

File: storage.bs [hunk 10]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+ <li>
+  <p>If <var>map</var>[<var>key</var>] does not <a for=map>exist</a>, then:
+
+  <ol>
+   <li><p>Let <var>unit</var> be a new <a>storage unit</a>.
+
+   <li><p>Set <var>unit</var>'s <a>bucket map</a>["<code>default</code>"] to the result of
+   <a>create a storage bucket</a> with <var>type</var>.
+
+   <li><p>Set <var>map</var>[<var>key</var>] to <var>unit</var>.
```
Comment (inexorabletash) [inline storage.bs:157]: This may be clarified when addressing domenic's comments, but: as written, this would imply that the first "obtain" call with a given _type_ populates the bucket map with a unit of that type. This is fine, since a bucket map's contents should all be of the same type. But if that's the case, perhaps the bucket map should have a type, rather than it being always passed in.

File: storage.bs [hunk 11]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<a for=/>storage map</a>.
+
+<p class=XXX>Browsing session is yet to be formally defined. For all intents and purposes it is a
+<a>top-level browsing context</a>, except that it survives the <a>top-level browsing context</a>
+being replaced due to a cross-origin opener policy.
+
+
+<h3 id=storage-endpoints>Storage endpoints</h3>
+
+<p>A <dfn export>storage endpoint</dfn> is a more-durable or session storage API that uses the
```
Comment (inexorabletash) [inline storage.bs:178]: Do you want to quote/style "more-durable" and "session" here, for consistency?

File: storage.bs [hunk 12]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<!-- Note, these will be exposed more generally through https://github.com/whatwg/storage/pull/69 so
+     please keep https://w3ctag.github.io/design-principles/#casing-rules in mind. -->
+
+<table>
+ <tr>
+  <th><a>Storage identifier</a>
+  <th><a>Storage types</a>
+ <tr>
+  <td>"<code>caches</code>"
+  <td>« "<code>more-durable</code>" »
```
Comment (inexorabletash) [inline storage.bs:204]: It was discussed above but: since all of these sets have exactly one member, I'd appreciate a non-normative note calling out that this is intentional.

File: storage.bs [hunk 13]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+ <tr>
+  <td>"<code>serviceWorkerRegistrations</code>"
+  <td>« "<code>more-durable</code>" »
+ <tr>
+  <td>"<code>sessionStorage</code>"
+  <td>« "<code>session</code>" »
+</table>
+
+<p class="note no-backref">Standards can use these <a>storage identifiers</a> with
+<a>obtain a more-durable storage bottle map</a> and <a>obtain a session storage bottle map</a>.
```
Comment (inexorabletash) [inline storage.bs:220]: https://youtu.be/mZYu7DnsCF0?t=3

File: storage.bs [hunk 14]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<p>A <dfn id=bucket oldids=box>more-durable storage bucket</dfn> is a <a>storage bucket</a> for
+more-durable storage APIs.
 
-<p>A <a>bucket</a> is considered to be an atomic unit. Whenever a <a>bucket</a> is cleared by the
-user agent, it must be cleared in its entirety.
+<p>A <a>more-durable storage bucket</a> has a
+<dfn for="more-durable storage bucket" id=bucket-mode oldids=box-mode>mode</dfn>, which is
+"<code>best-effort</code>" or "<code>persistent</code>". It is initially "<code>best-effort</code>".
+
+<hr>
```
Comment (inexorabletash) [inline storage.bs:252]: FWIW, this renders as a substantial amount of whitespace. I thought something was wrong with the spec, e.g. a missing image.

File: storage.bs [hunk 15]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<p>To <dfn>create a storage bucket</dfn>, given a <a>storage type</a> <var>type</var>, run these
+steps:
+
+<ol>
+ <li><p>Let <var>bucket</var> be null.
+
+ <li><p>If <var>type</var> is "<code>more-durable</code>", then set <var>bucket</var> to a new
+ <a>more-durable storage bucket</a>.
+
+ <li><p>Otherwise, if <var>type</var> is "<code>session</code>", set <var>bucket</var> to a new
```
Comment (inexorabletash) [inline storage.bs:267]: nit: add "then" (I think we try and use if/then consistently?)

File: storage.bs [hunk 16]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+ <a>session storage bucket</a>.
+
+ <li><p>Assert: <var>bucket</var> is a <a>storage bucket</a>.
```
Comment (inexorabletash) [inline storage.bs:270]: Aside: this Assert surprised me, since it's not validating a precondition for the steps or some other algorithm, until I realized it's due to the "otherwise" case having an "if" and trying to ensure there's no new type snuck in to the enumeration? Or alternately write e.g.  Maybe a "Switch on..." would be clearer?

I've also written this as _Otherwise (type is session), set ..._, which is admittedly an implicit assertion tucked in there.

File: storage.bs [hunk 17]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+  <var>environment</var>'s <span class=XXX>browsing session</span>'s
+  <a for="browsing session">storage map</a>.
+
+  <p class="XXX">See
+  <a href="https://github.com/whatwg/html/issues/4782">whatwg/html issue #4782</a> and
+  <a href="https://github.com/whatwg/html/issues/5350">whatwg/html issue #5350</a> for defining
+  browsing session. It is roughly analogous to <a>top-level browsing context</a> except that it
+  cannot be replaced due to <code>Cross-Origin-Opener-Policy</code> or navigation.
+
+ <li><p>Assert: <var>map</var> is a <a for=/>storage map</a>.
```
Comment (inexorabletash) [inline storage.bs:311]: Same aside as above.

File: storage.bs [hunk 18]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<a>environment settings object</a> <var>environment</var>, and a <a>storage type</a>
+<var>type</var>, run these steps:
+
+<ol>
+ <li><p>Let <var>key</var> be <var>environment</var>'s
+ <a for="environment settings object">origin</a>.
+
+ <li><p>If <var>key</var> is an <a>opaque origin</a>, then return failure.
+
+ <li><p>If the user has disabled storage, then return failure.
```
[the remark this answers is not in the document]
Comment (annevk) [reply storage.bs:146]: Maybe as a follow-up? I'm not entirely sure what that would look like, but happy for someone to take a stab at it.

File: storage.bs [hunk 19]
```diff
@@ -94,37 +100,252 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<p>This standard primarily concerns itself with storage.
+
+
+
+<h2 id=model>Model</h2>
+
+
+<h3 id=storage-units>Storage units</h3>
+
+<p>A <dfn>storage type</dfn> is "<code>more-durable</code>" or "<code>session</code>".
```
[the remark this answers is not in the document]
Comment (annevk) [reply storage.bs:117]: Well, it's not permanent. I considered "temporal" and "pressure-based", but maybe "local" after `localStorage` is good enough after all and from the table it's clear it applies to other endpoints too. I'm going with that.

File: storage.bs (hunk 11 above)
Comment (annevk) [reply storage.bs:178]: Here they are used in an informal manner to describe the APIs. Elsewhere they're a type value.

File: storage.bs (hunk 12 above)
Comment (annevk) [reply storage.bs:204]: The source note is not sufficient?

File: storage.bs (hunk 14 above)
Comment (annevk) [reply storage.bs:252]: Perhaps worth reporting at https://github.com/whatwg/whatwg.org/. We use this pattern all over.

File: storage.bs (hunk 15 above)
Comment (annevk) [reply storage.bs:267]: We don't use it after otherwise.

File: storage.bs (hunk 16 above)
Comment (annevk) [reply storage.bs:270]: This was intended for absolute clarity. If you want I could put the assert inside the otherwise branch and make that have substeps.

File: storage.bs (hunk 10 above)
Comment (annevk) [reply storage.bs:157]: The type only matters all the way down so I'd prefer keeping the more generic structures free of them.

File: storage.bs (hunk 12 above)
Comment (annevk) [reply storage.bs]: Ended up describing this anyway.

**Comment by annevk** [comment]
I did the refactoring in batches, but reviewing https://whatpr.org/storage/86.html#model is probably easier. Note that apart from the Model section this also improved various aspects of the Storage Standard itself, by making it much more clear what exactly various aspects are talking about and when things might fail and such.

File: storage.bs [hunk 20]
```diff
@@ -94,37 +100,292 @@ function retrieveNextChunk(nextChunkInfo) {
...
+<a href="https://github.com/whatwg/html/issues/4782">whatwg/html issue #4782</a> and
+<a href="https://github.com/whatwg/html/issues/5350">whatwg/html issue #5350</a> for defining
+browsing session. It is roughly analogous to <a>top-level browsing context</a> except that it cannot
+be replaced due to <code>Cross-Origin-Opener-Policy</code> or navigation.
+
+
+<h3 id=storage-shelves>Storage shelves</h3>
+
+<p>A <dfn oldids=site-storage-unit lt="storage shelf|storage shelves">storage shelf</dfn> is the
+biggest storage container, apart from the user agent itself. It holds a <dfn>bucket map</dfn>, which
```
Comment (inexorabletash) [inline storage.bs:216]: Isn't a shed bigger than a shelf? (I'm not sure of the utility of calling out "biggest storage container" here)

Comment (asutherland) [reply storage.bs:216]: I also find this bit a little odd and after the terminology change probably no longer necessary to clarify the sizes beyond the hierarchical containment relationship.

Right now it's:
> A storage shelf is the biggest storage container, apart from the user agent itself. It holds a bucket map, which is a map of strings to storage buckets. 

And maybe it could instead be something like:
> **A storage shelf is the unit of per-origin data storage within a storage shed.**  It holds a bucket map, which is a map of strings to storage buckets. 

Alternately:
> **A storage shelf exists for each origin within a storage shed.  Each storage shelf holds a bucket map that maps strings to storage buckets.**

File: storage.bs [hunk 21]
```diff
@@ -94,37 +100,292 @@ function retrieveNextChunk(nextChunkInfo) {
...
+will give them:
+
+<ul>
+ <li><p>Failure, which might mean the API has to throw or otherwise indicate there is no storage
+ available for that <a for=/>environment settings object</a>.
+
+ <li><p>A <a>storage proxy map</a> that operates analogously to a <a for=/>map</a>, which can be
+ used to store data in a manner that suits the API. This standard takes care of isolating that data
+ appropriately within a user agent or browsing session. Isolating it from other <a for=/>origins</a>
+ as well as other APIs.
```
Comment (domenic) [inline storage.bs:126]: This is not a complete sentence.

File: storage.bs [hunk 22]
```diff
@@ -94,37 +100,292 @@ function retrieveNextChunk(nextChunkInfo) {
...
+</ul>
+
+<p class=note>If you are defining a standard for such an API, consider filing an issue agains this
+standard for assistence and review.
```
Comment (domenic) [inline storage.bs:130]:
```suggestion
standard for assistance and review.
```

File: storage.bs [hunk 23]
```diff
@@ -94,37 +100,292 @@ function retrieveNextChunk(nextChunkInfo) {
...
-<a for=bucket>mode</a> is <em>not</em> "<code title>persistent</code>".
+<p>A <a>local storage bucket</a> has a
+<dfn for="local storage bucket" id=bucket-mode oldids=box-mode>mode</dfn>, which is
+"<code>best-effort</code>" or "<code>persistent</code>". It is initially "<code>best-effort</code>".
 
-<p>A <a>bucket</a> is considered to be an atomic unit. Whenever a <a>bucket</a> is cleared by the
-user agent, it must be cleared in its entirety.
+<hr>
+
+<p>A <dfn>session storage bucket</dfn> is a <a>storage bucket</a> for session storage APIs.
```
Comment (domenic) [inline storage.bs:277]: Can we have definitions and cross-references for "local storage" and "session storage"? Maybe

> A user agent holds a storage shed, which is a storage shed. All **local storage** data will be stored in this shed.

File: storage.bs [hunk 24]
```diff
@@ -94,37 +100,292 @@ function retrieveNextChunk(nextChunkInfo) {
...
+
+
+<h3 id=storage-proxy-maps>Storage proxy maps</h3>
+
+<p>A <dfn>storage proxy map</dfn> is equivalent to a <a for=/>map</a>, except that all operations
+are instead performed on its <dfn for="storage proxy map">backing map</dfn>.
+
+<p class="XXX">This allows for the <a for="storage proxy map">backing map</a> to be replaced. This
+is needed for <a href="https://github.com/whatwg/storage/issues/4">issue #4</a> and potentially the
+<a href="https://privacycg.github.io/storage-access/">Storage Access API</a>.
```
Comment (domenic) [inline storage.bs:380]: I guess maybe it'll be used as part of the future infrastructure, but right now the "proxy map reference set" is append-only, and never read from.

Comment (asutherland) [reply storage.bs:380]: I also share this concern, although it seems like that's a future-work follow-up on the "replace" algorithm proposed at the bottom of https://github.com/whatwg/storage/issues/18#issuecomment-614751336 ?

I'm assuming that's what's going to happen, in which case we might as well wait, as it seems likely to be non-trivial.

Comment (annevk) [reply storage.bs:380]: Yeah you need to know who have a proxy map in order to let their proxies point elsewhere when storage is replaced.

Comment (annevk) [reply storage.bs:380]: I'm a bit surprised by this feedback though as I thought from #18 that we agreed on this approach.
